### ISSUE TITLE
Unexpected Behavior With table.Copy

### ISSUE THREAD

**Issue by shadowscion** [Facepunch/garrysmod-issues#2882]
### Details
**table.Copy** doesn't create copies of vectors or angles within the table. I'm not sure if this is intended, but to me, it's unexpected. It seems that vectors and angles are not treated as tables by the **istable** function that is used in table.Copy.

### Steps to reproduce
```
local original = {
    ang = Angle(0, 0, 0),
    pos = Vector(1, 1, 1),
    dir = Vector(1, 0, 0),
}

MsgC(Color(255, 255, 255), "\n\nOriginal Before\n")
PrintTable(original)



local copy = table.Copy(original)

copy.ang:RotateAroundAxis(copy.dir, 90)
copy.pos:Rotate(copy.ang)




MsgC(Color(255, 255, 255), "Original After\n")
PrintTable(original)

MsgC(Color(255, 255, 255), "Copy\n")
PrintTable(copy)



Output...

Original Before
ang =   0.000 0.000 0.000
dir =   1.000000 0.000000 0.000000
pos =   1.000000 1.000000 1.000000
Original After
ang =   -0.000 0.000 90.000
dir =   1.000000 0.000000 0.000000
pos =   1.000000 -1.000000 1.000000
Copy
ang =   -0.000 0.000 90.000
dir =   1.000000 0.000000 0.000000
pos =   1.000000 -1.000000 1.000000
```

**Comment by Jvs34** [comment]
Doesn't seem unexpected to me, they're not a table, they're userdata.
Colors however are actual tables, not userdata.
I understand how this is a pain to handle but that's how things are and I don't see how this would be fixed without some hacky workaround specifically for those two data types.

**Comment by robotboy655** [comment]
table.Copy doesn't actually copy any userdata. So yeah, this is to be expected.

**Comment by shadowscion** [comment]
I understand, and that's why I didn't call it a bug. But it was a complete nightmare to trace the source of the problems I was encountering to this.

**Comment by robotboy655** [comment]
Updated the wiki page. For now, you can rename & edit the function and manually copy the userdata you need.

**Comment by shadowscion** [comment]
@robotboy655 Thanks.

**Comment by Acecool** [comment]
Typically a table.Copy is a "shallow-copy" which would copy the simple data-type data and references to other data-types... You could create a DeepCopy ( although one may already exist?? ) to handle this.

**Comment by bmwalters** [comment]
@Acecool 
> Typically a table.Copy is a "shallow-copy" which would copy the simple data-type data and references to other data-types... You could create a DeepCopy ( although one may already exist?? ) to handle this.

False. table.Copy *is* a deep copy function. To test this yourself:
```lua
local a = { test = { "Hello World" } }
print(table.Copy(a).test == a.test) -- false
```

The only problem is *userdata* copying is not handled; to handle copying userdata inside table.Copy you would have to have a special case for each type of value. A better solution would be requiring a `:Copy()` method on each GMod userdata (comparable to the copy constructor in C++), and table.Copy could just call that method.

**Comment by Acecool** [comment]
I said "typically" it is a shallow copy... If GMod table.Copy is supposed to be a deep-copy then the data should be cloned...

Why not submit a pull-request with Copy / Clone added for each of the meta-tables? I tried adding Serialize to fix a game-breaking bug but I don't know if that ever got merged... Sometimes things do get merged, other times not...
